src/main/java/uk/gov/hmcts/darts/courthouse/service/CourthouseServiceImpl.java

```diff
@@ -285,11 +297,66 @@ public AdminCourthouse updateCourthouse(Integer courthouseId, CourthousePatch co
+        
+        var updateDiffsForAuditing = updateDiffsOf(courthouseEntity, courthousePatch);
 
         var patchedCourthouse = courthouseUpdateMapper.mapPatchToEntity(courthousePatch, courthouseEntity);
         courthouseRepository.saveAndFlush(patchedCourthouse);
 
+        auditUpdates(updateDiffsForAuditing);
+
         return courthouseUpdateMapper.mapEntityToAdminCourthouse(patchedCourthouse);
     }
 
+    private void auditUpdates(Map<AuditActivity, Boolean> updateDiffsForAuditing) {
+        if (updateDiffsForAuditing.get(UPDATE_COURTHOUSE_GROUP)) {
+            auditApi.recordAudit(UPDATE_COURTHOUSE_GROUP, authorisationApi.getCurrentUser(), null);
+        }
+
+        if (updateDiffsForAuditing.get(UPDATE_COURTHOUSE)) {
+            auditApi.recordAudit(UPDATE_COURTHOUSE, authorisationApi.getCurrentUser(), null);
+        }
+    }
+
+    private Map<AuditActivity, Boolean> updateDiffsOf(CourthouseEntity courthouseEntity, CourthousePatch courthousePatch) {
+        Map<AuditActivity, Boolean> updateDiffs = new HashMap<>();
+
+        if (isBasicDetailsUpdated(courthousePatch, courthouseEntity)) {
+            updateDiffs.put(AuditActivity.UPDATE_COURTHOUSE, true);
+        } else {
+            updateDiffs.put(AuditActivity.UPDATE_COURTHOUSE, false);
+        }
+        if (securityGroupsUpdated(courthousePatch, courthouseEntity)) {
+            updateDiffs.put(AuditActivity.UPDATE_COURTHOUSE_GROUP, true);
+        } else {
+            updateDiffs.put(AuditActivity.UPDATE_COURTHOUSE_GROUP, false);
+        }
+        return updateDiffs;
+    }
+
+    private static boolean isBasicDetailsUpdated(CourthousePatch patch, CourthouseEntity prePatched) {
+        boolean isNameUpdated = isNotNullAndDifferent(patch.getCourthouseName(), prePatched.getCourthouseName());
+        boolean isDisplayNameUpdated = isNotNullAndDifferent(patch.getDisplayName(), prePatched.getDisplayName());
+        boolean isRegionUpdated = isRegionUpdated(patch, prePatched);
+
+        return isNameUpdated || isDisplayNameUpdated || isRegionUpdated;
+    }
+
+    private static boolean isNotNullAndDifferent(String patchValue, String prePatchedValue) {
+        return nonNull(patchValue) && !patchValue.equals(prePatchedValue);
+    }
+
+    private static boolean isRegionUpdated(CourthousePatch patch, CourthouseEntity prePatched) {
+        return nonNull(patch.getRegionId()) && (prePatched.getRegion() == null || !patch.getRegionId().equals(prePatched.getRegion().getId()));
+    }
+
+    private static boolean securityGroupsUpdated(CourthousePatch courthousePatch, CourthouseEntity prePatchedEntity) {
+        if (courthousePatch.getSecurityGroupIds() == null) {
+            return false;
+        }
+        var prePatchedGroups = prePatchedEntity.getSecurityGroups().stream()
+            .map(SecurityGroupEntity::getId)
+            .toList();
+        return !prePatchedGroups.equals(courthousePatch.getSecurityGroupIds());
+    }
```

Comment:
I might be missing something but I wonder whether we need to write this custom audit logic. 
The power of Envers is that it provides automatic auditing by just adding annotations. I would have thought that we could cover the basic admin court case operations requirements we have by using only annotations. 
And then use the Envers API to figure out what type of changes where made to the entity when the need arise to access those changes.

Reply:
This isn't related to envers, we have custom auditing also.

Reply:
I believe the requirement is to record that something happened in the 'audit' table and then if they want more detail about what actually changed, they can query the envers table.